Update file_downloader.py

zengwei2000 opened the pull request:
Using the Python 3 backoff library to increase network error retries

zengwei2000 commented on mock/py/mockbuild/file_downloader.py:
```diff
@@ -4,6 +4,7 @@
 import cgi
 import shutil
 import tempfile
+import backoff
```
Okay, could you please check if this PR meets the requirements

praiskup replied:
Yes!  Just a nit, the non-stdlib imports should be imported later (somewhere along python-requests).

praiskup commented on mock/py/mockbuild/file_downloader.py:
```diff
@@ -61,7 +63,7 @@ def get(cls, pkg_url_or_local_file):
             cls.backmap[pkg] = url
             return pkg
-        except Exception as err:  # pylint: disable=broad-except
+        except requests.exceptions.RequestException as err:
```
I'm not sure this will work, though.  If you catch the exception, the wrapping decorator can not react on it anymore.

zengwei2000 replied:
I added a handle for all exceptions, I'm not sure if this is possible

praiskup replied:
Thank you for the quick update!

I think there's still the same issue.  When we do `except Exception` inside, the exception is simply swallowed and the external logic (the backoff decorator) can not see it to react appropriately.  I think we could go with two methods, e.g. keep the wrapping `get()` method that could call a backoff-decorated `_get_inner()` method.

zengwei2000 replied:
Thank you for pointing out this issue. I have made modifications to it and I am not sure if it is correct

praiskup commented on mock/py/mockbuild/file_downloader.py:
```diff
@@ -42,28 +43,33 @@ def get(cls, pkg_url_or_local_file):
+        req = requests.get(url)
+        req.raise_for_status()
+        if req.status_code != requests.codes.ok:
+            log.error("Can't download {}".format(url))
```
Do you want to just fail here, or should we retry for some GET errors?

praiskup commented on mock/py/mockbuild/file_downloader.py:
```diff
@@ -42,28 +43,33 @@ def get(cls, pkg_url_or_local_file):
-            if req.status_code != requests.codes.ok:
-                log.error("Can't download {}".format(url))
-                return False
+            return cls._get_inner(pkg)
```
I'd move the `log.info('Fetching remote file %s', url)` before this line.

praiskup commented on mock/py/mockbuild/file_downloader.py:
```diff
@@ -42,28 +43,33 @@
+        req = requests.get(url)
+        req.raise_for_status()
+        if req.status_code != requests.codes.ok:
+            log.error("Can't download {}".format(url))
```
This is old thing, detected as new because the method name is different now.  Would you mind changing it to `log.error("Can't download %s", url)`?

praiskup commented:
We need to have `Requires: python3-backoff` in `mock.spec`.

zengwei2000 commented:
> We need to have `Requires: python3-backoff` in `mock.spec`.

https://github.com/rpm-software-management/mock/pull/1131   I have modified my submission according to your request and rewritten the commit information
